File: src/main/java/ch/obermuhlner/astro/gradient/filter/Filter.java
package ch.obermuhlner.astro.gradient.filter;

import ch.obermuhlner.astro.image.ArrayDoubleImage;
import ch.obermuhlner.astro.image.DoubleImage;
import ch.obermuhlner.astro.image.SubDoubleImage;

public interface Filter {
  DoubleImage filter(DoubleImage source, DoubleImage target, int width, int height);

  default DoubleImage filter(DoubleImage source, DoubleImage target) {
    int width = Math.min(source.getWidth(), target.getWidth());
    int height = Math.min(source.getHeight(), target.getHeight());
    return filter(source, target, width, height);
  }

  default DoubleImage filter(DoubleImage source) {
    ArrayDoubleImage target = new ArrayDoubleImage(source.getWidth(), source.getHeight(), source.getColorModel());
    return filter(source, target);
  }

  default DoubleImage filter(DoubleImage source, int sourceX, int sourceY, DoubleImage target, int targetX, int targetY, int width, int height) {
    SubDoubleImage subSource = new SubDoubleImage(source, sourceX, sourceY, width, height);
    SubDoubleImage subTarget = new SubDoubleImage(target, targetX, targetY, width, height);

    return filter(subSource, subTarget);
  }

}

File: src/main/java/ch/obermuhlner/astro/gradient/filter/Filter.kt
package ch.obermuhlner.astro.gradient.filter

import ch.obermuhlner.astro.image.ArrayDoubleImage
import ch.obermuhlner.astro.image.DoubleImage
import kotlin.math.min

open interface Filter {
    fun filter(source: DoubleImage, target: DoubleImage, width: Int, height: Int): DoubleImage

    fun filter(source: DoubleImage, target: DoubleImage): DoubleImage {
        val width: Int = min(source.width, target.width)
        val height: Int = min(source.height, target.height)
        return filter(source, target, width, height)
    }

    fun filter(source: DoubleImage): DoubleImage {
        val target = ArrayDoubleImage(source.width, source.height, source.colorModel)
        return filter(source, target)
    }

    fun filter(source: DoubleImage, sourceX: Int, sourceY: Int, target: DoubleImage, targetX: Int, targetY: Int, width: Int, height: Int): DoubleImage {
        val subSource = source.croppedImage(sourceX, sourceY, width, height, false)
        val subTarget = target.croppedImage(targetX, targetY, width, height, false)
        return filter(subSource, subTarget)
    }
}